File: task-03/Go/Sub-Task4/code4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func generateDiamond(n int) []string {
	var diamond []string

	// Upper part of the diamond
	for i := 1; i <= n; i++ {
		spaces := strings.Repeat(" ", n-i)
		stars := strings.Repeat("*", 2*i-1)
		diamond = append(diamond, spaces+stars)
	}

	// Lower part of the diamond
	for i := n - 1; i >= 1; i-- {
		spaces := strings.Repeat(" ", n-i)
		stars := strings.Repeat("*", 2*i-1)
		diamond = append(diamond, spaces+stars)
	}

	return diamond
}

func writeToFile(n int) {
	diamond := generateDiamond(n)
	file, err := os.Create("output.txt")
	if err != nil {
		fmt.Println("Error creating file:", err)
		return
	}
	defer file.Close()

	writer := bufio.NewWriter(file)
	for _, line := range diamond {
		_, err := writer.WriteString(line + "\n")
		if err != nil {
			fmt.Println("Error writing to file:", err)
			return
		}
	}
	writer.Flush()
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error opening input file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	n, err := strconv.Atoi(scanner.Text())
	if err != nil {
		fmt.Println("Error converting input to integer:", err)
		return
	}

	writeToFile(n)
	fmt.Println("Diamond pattern generated in output.txt")
}

File: task-03/C++/Sub-Task4/code4.cpp
#include <iostream>
#include <fstream>

void generateDiamond(int n, std::ofstream &fout) {
    // Upper part of the diamond
    for (int i = 1; i <= n; ++i) {
        for (int j = i; j < n; ++j) {
            fout << " ";
        }
        for (int j = 1; j <= (2 * i - 1); ++j) {
            fout << "*";
        }
        fout << "\n";
    }

    // Lower part of the diamond
    for (int i = n - 1; i >= 1; --i) {
        for (int j = n; j > i; --j) {
            fout << " ";
        }
        for (int j = 1; j <= (2 * i - 1); ++j) {
            fout << "*";
        }
        fout << "\n";
    }
}

int main() {
    int n;
    std::ifstream fin("input.txt");
    std::ofstream fout("output.txt");

    if (!fin.is_open()) {
        std::cerr << "Error opening input file!\n";
        return 1;
    }

    fin >> n;
    fin.close();

    generateDiamond(n, fout);
    fout.close();

    std::cout << "Diamond pattern generated in output.txt\n";
    return 0;
}

